### ISSUE TITLE
OneDrive for business: HTTP request returned status code 404 (Not Found)

### ISSUE THREAD

**Issue by lorenzosu** [abraunegg/onedrive#33]
Since this morning (but the file was actually created on Friday), onedrive --monitor crashes with a series of messages like the following:
```
Uploading file [FILEPATH]... HTTP request returned status code 404 (Not Found)
{
    "error": {
        "code": "itemNotFound",
        "innerError": {
            "date": "2018-06-25T13:15:08",
            "request-id": "b2a427bb-7028-4510-9e2e-abbb84540a33"
        },
        "message": "The resource could not be found."
    }
}
```

And then:
```----------------
??:? void onedrive.OneDriveApi.checkHttpCode(ref const(std.json.JSONValue)) [0xddbba4c0]
??:? std.json.JSONValue onedrive.OneDriveApi.upload(immutable(char)[], immutable(char)[]) [0xddbb9cc5]
??:? std.json.JSONValue onedrive.OneDriveApi.simpleUploadReplace(immutable(char)[], immutable(char)[], immutable(char)[], const(char)[]) [0xddbb8a16]
??:? void sync.SyncEngine.uploadFileDifferences(itemdb.Item, immutable(char)[]) [0xddbc4fa4]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc490d]
??:? void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0xddbc4acd]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc48da]
??:? void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0xddbc4acd]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc48da]
??:? void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0xddbc4acd]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc48da]
??:? void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0xddbc4acd]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc48da]
??:? void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0xddbc4acd]
??:? void sync.SyncEngine.uploadDifferences(itemdb.Item) [0xddbc48da]
??:? void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0xddbc4736]
??:? void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool) [0xddbab6ab]
??:? _Dmain [0xddbab1c0]
```

**Comment by lorenzosu** [comment]
Same if I do:

`onedrive --synchronize --verbose`

**Comment by lorenzosu** [comment]
Update: I temporarily moved the (local) single file which was triggering this, did a sync and other files synced successfully. Then moved that file back again to the original (local) location and starting the monitor again synced it correctly.

So there seem to be 2 issues:
1. why is onedrive client triggering a 404 error for a file in the first place?
2. that error blocks sync of any other file

**Comment by abraunegg** [comment]
Firstly - are you running the latest code - as you have not indicated this in your post. There were some updates / tweaks for another issue (#11, #17) where some additional checks were added. Can you get the latest code and re-test / reproduce this?

Why it is generating a 404 ? Unsure - is there anything unique about the filename?

Now - given that this is also a Business Account, and that it is failing with in this function - `simpleUploadReplace` you are most likely being hit by 2 side issues:

* #23 - Where any file larger than 64Kb generates an incorrect hash. This has been the case since 'skilion' implemented QuickXorHash to support OneDrive Business. It means that even if you just 'touch' a file, the code will think that the contents of the file has changed (when it has not) and re-upload the file.
* #2 - Where uploaded files generate 2 versions for every upload. I have this "mostly" fixed, however there is an API Blocker where uploading a modified file to replace a file on OneDrive still creates 2 new versions.

I cant say if this is a bug at the moment - but if you can re-test with the latest code & see if you can reproduce that would be greatly appreciated. If you can, can you post a full verbose log.

**Comment by lorenzosu** [comment]
Yes I was running the latest code cloned just before the bug report.

Nothing particular with the file that I can recall of. I might have been offline when saving it to the directory, but couldn't confirm 100%

Unfortunately I can't re-test as I 'restored' the file and now its' working. Any way to try and simulate this? Also, should this happen again?

Finally, but I don't think it's really relevant, I use onedrive with proxychains when in the office as we have a corporate proxy. Up to now this hasn't created any problem.

**Comment by lorenzosu** [comment]
Ok, one small update, not sure if it is relevant, but it may as you meantioned the hash.In the same directory of the 'offending' file I had created a copy (with different name) with the intent of updating it (it's a presentation file), and then didn't get round to doing it. So essentially you have in the same directory two identical files which only have a different file name and timestamp.

**Comment by abraunegg** [comment]
Thanks .. I think what is happening here is the 404 error potentially was caused by #23, the fix is both #23 and #2. Whilst I am working on trying to get the hashing sorted, I cant fix the OneDrive API bug that's blocking #2 from being completed.

Once #23 is resolved - the 'chances' of having to re-upload a file when only the timestamp has changed will drop significantly, thus significantly reduce the possibility of the error. #2 changes the upload method as well for OneDrive Business & this will close this out.

**Comment by abraunegg** [comment]
As this issue has 'technically' been resolved at your end, but you are side impacted by #2 & #23, will close this case for now.
